src/cwts/networkanalysis/CPMClusteringAlgorithm.java

```diff
@@ -135,4 +137,154 @@ public double calcQuality(Network network, Clustering clustering)
+        double[] clusterWeights, totalEdgeWeightPerCluster;
+        int i, j;
+
+        clusterWeights = new double[clustering.nClusters];
+        totalEdgeWeightPerCluster = new double[clustering.nClusters];
+        for (i = 0; i < network.nNodes; i++)
+        {
+            clusterWeights[clustering.clusters[i]] += network.nodeWeights[i];
+            if (clustering.clusters[i] == cluster)
+                for (j = network.firstNeighborIndices[i]; j < network.firstNeighborIndices[i + 1]; j++)
+                    totalEdgeWeightPerCluster[clustering.clusters[network.neighbors[j]]] += network.edgeWeights[j];
+        }
+
+        i = -1;
+        maxQualityFunction = 0;
+        for (j = 0; j < clustering.nClusters; j++)
+            if ((j != cluster) && (clusterWeights[j] > 0))
+            {
+                qualityFunction = totalEdgeWeightPerCluster[j] / clusterWeights[j];
+                if (qualityFunction > maxQualityFunction)
+                {
+                    i = j;
+                    maxQualityFunction = qualityFunction;
+                }
+            }
+
+        if (i >= 0)
+        {
+            for (j = 0; j < network.nNodes; j++)
+                if (clustering.clusters[j] == cluster)
+                    clustering.clusters[j] = i;
+            if (cluster == clustering.nClusters - 1)
+                clustering.nClusters = Arrays.calcMaximum(clustering.clusters) + 1;
+        }
+
+        return i;
+    }
+
+    /**
+     * Removes small clusters from a clustering. Clusters are merged until each
+     * cluster contains at least a certain minimum number of nodes.
+     * 
+     * @param network             Network
+     * @param clustering          Clustering
+     * @param minNNodesPerCluster Minimum number of nodes per cluster
+     * 
+     * @return Boolean indicating whether any clusters have been removed
+     */
+    public boolean removeSmallClustersBasedOnNNodes(Network network, Clustering clustering, int minNNodesPerCluster)
+    {
+        int i, j, nNodesSmallestCluster;
+        int[] nNodesPerCluster;
+
+        Network reducedNetwork = network.createReducedNetwork(clustering);
+        Clustering clusteringReducedNetwork = new Clustering(reducedNetwork.nNodes);
+
+        nNodesPerCluster = clustering.getNNodesPerCluster();
+
+        do
+        {
+            i = -1;
+            nNodesSmallestCluster = minNNodesPerCluster;
+            for (j = 0; j < clustering.nClusters; j++)
+                if ((nNodesPerCluster[j] > 0) && (nNodesPerCluster[j] < nNodesSmallestCluster))
+                {
+                    i = j;
+                    nNodesSmallestCluster = nNodesPerCluster[j];
+                }
+
+            if (i >= 0)
+            {
+                j = removeCluster(reducedNetwork, clusteringReducedNetwork, i);
+                if (j >= 0)
+                    nNodesPerCluster[j] += nNodesPerCluster[i];
+                nNodesPerCluster[i] = 0;
```

Comment:
Just to be sure: should we not throw an exception if something like this happens? If, for whatever reason, this occurs (e.g. a user supplied membership), it might cause a rather hard to track down bug.